src/ASM/LR/ASMu2D.C

```diff
@@ -1288,60 +1237,49 @@ bool ASMu2D::integrate (Integrand& integrand,
 
       // --- Integration loop over all Gauss points in each direction ----------
 
-      int jp = (iel-1)*nGP*nGP;
-      fe.iGP = firstIp + jp; // Global integration point counter
+      size_t ip = 0;
+      fe.iGP = firstIp + (iel-1)*ng[0]*ng[1]; // Global integration point counter
 
-      for (int j = 0; j < nGP; j++)
-        for (int i = 0; i < nGP; i++, fe.iGP++)
+      for (int j = 0; j < ng[1]; j++)
+        for (int i = 0; i < ng[0]; i++, fe.iGP++, ++ip)
         {
           // Local element coordinates of current integration point
-          fe.xi  = xg[i];
-          fe.eta = xg[j];
+          fe.xi  = xg[0][i];
+          fe.eta = xg[1][j];
 
           // Parameter values of current integration point
-          fe.u = param[0] = gpar[0][i];
-          fe.v = param[1] = gpar[1][j];
+          fe.u = param[0] = cache.getParam(0,iel-1,i);
+          fe.v = param[1] = cache.getParam(1,iel-1,j);
 
-          // Extract basis function derivatives at current integration point
-          if (use3rdDer)
-            SplineUtils::extractBasis(spline3[fe.iGP-firstIp],fe.N,dNdu,d2Ndu2,d3Ndu3);
-          else if (use2ndDer)
-            SplineUtils::extractBasis(spline2[fe.iGP-firstIp],fe.N,dNdu,d2Ndu2);
-          else {
-            SplineUtils::extractBasis(spline1[fe.iGP-firstIp],fe.N,dNdu);
+          const BasisFunctionVals& bfs = cache.getVals(iel-1,ip);
+          fe.N = bfs.N;
 #if SP_DEBUG > 4
-            if (iel == dbgElm || iel == -dbgElm || dbgElm == 0)
-            {
-              const Go::BasisDerivsSf& spline = spline1[fe.iGP-firstIp];
-              std::cout <<"\nBasis functions at a integration point "
-                        <<" : (u,v) = "<< spline.param[0] <<" "<< spline.param[1]
-                        <<"  left_idx = "<< spline.left_idx[0]
-                        <<" "<< spline.left_idx[1];
-              for (size_t ii = 0; ii < spline.basisValues.size(); ii++)
-                std::cout <<'\n'<< 1+ii <<'\t' << spline.basisValues[ii] <<'\t'
-                          << spline.basisDerivs_u[ii] <<'\t'
-                          << spline.basisDerivs_v[ii];
-              std::cout << std::endl;
-            }
-#endif
+          if (iel == dbgElm || iel == -dbgElm || dbgElm == 0)
+          {
+            std::cout <<"\nBasis functions at a integration point "
+                      <<" : (u,v) = "<< fe.u <<" "<< fe.v;
+            for (size_t ii = 0; ii < bfs.N.size(); ii++)
+              std::cout <<'\n'<< 1+ii <<'\t' << bfs.N[ii] <<'\t'
```

Comment:
A missing semicolon here causes compilation failure when SP_DEBUG > 4.
See fix in the last commit on my branch.

Reply:
picked.